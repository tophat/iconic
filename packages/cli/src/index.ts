import * as path from 'path'

import yargs from 'yargs'
import * as fs from 'fs-extra'
import convert from '@react-vector-graphics/core'

function run(src: string, dest: string): void {
    const files = fs.readdirSync(src)
    const filesContents = files.map((file): [string, string] => [
        file,
        fs.readFileSync(path.join(src, file), { encoding: 'utf-8' }),
    ])
    const results = filesContents.map(([file, contents]): object =>
        convert(file, contents),
    )
    for (const result of results) {
        for (const [file, contents] of Object.entries(result)) {
            const filePath = path.join(dest, file)
            fs.outputFileSync(filePath, contents)
        }
    }
}

const argv = yargs
    .usage('Usage: $0 -s [src] -d [dest]')
    .option('src', {
        alias: 's',
        describe: 'source folder',
        type: 'string',
    })
    .option('dest', {
        alias: 'd',
        describe: 'destination folder',
        type: 'string',
    })
    .demandOption(['s', 'd']).argv

const src = argv.s as string
const dest = argv.d as string
run(src, dest)
