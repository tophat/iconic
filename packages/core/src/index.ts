export { getLogger } from './logging'
export { run as default } from './main'
