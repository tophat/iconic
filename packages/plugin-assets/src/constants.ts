const ns = (s: TemplateStringsArray): string => `assets/${s}`

export const OPTIONS = {
    FILE_EXT: ns`fileExt`,
    GLOB_PATTERN: ns`globPattern`,
    NAME_SCHEME: ns`nameScheme`,
    OUTPUT_PATH: ns`outputPath`,
}

export const STATE = {
    COMPONENT_FILES: ns`componentFiles`,
    COMPONENT_NAME: 'componentName', // @svgr/plugin-jsx
    FILE_PATH: 'filePath', // @svgr/plugin-jsx
}
