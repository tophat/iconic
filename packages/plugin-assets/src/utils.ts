import * as path from 'path'

import { NamingScheme } from '@react-vector-graphics/types'

const capitalize = (w: string): string =>
    `${w[0].toUpperCase()}${w.slice(1).toLowerCase()}`

export const pathToName = (
    filePath: string,
    namingScheme: NamingScheme,
): string | undefined => {
    const words = path
        .basename(filePath, filePath.split('.').pop())
        .match(/[a-zA-Z0-9]+/g)
    switch (namingScheme) {
        case 'CONSTANT_CASE':
            return words?.join('_').toUpperCase()
        case 'snake_case':
            return words?.join('_').toLowerCase()
        case 'spinal-case':
            return words?.join('-').toLowerCase()
        case 'camelCase':
            return words
                ?.map((w, i): string => {
                    return i > 0 ? capitalize(w) : w.toLowerCase()
                })
                .join('')
        case 'PascalCase':
        default:
            return words?.map(capitalize).join('')
    }
}
