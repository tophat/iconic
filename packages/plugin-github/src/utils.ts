import { Logger } from '@react-vector-graphics/types'

export const fromBase64 = (data: string): string =>
    Buffer.from(data, 'base64').toString()

export const toBase64 = (data: string): string =>
    Buffer.from(data).toString('base64')

export const eagerPromises = (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    promises: Promise<any>[],
    logger: Logger = console,
): // eslint-disable-next-line @typescript-eslint/no-explicit-any
Promise<any> => Promise.all(promises.map(p => p.catch(logger.error)))
