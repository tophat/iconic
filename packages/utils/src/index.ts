export { NAMING_SCHEME, pathToName } from './pathToName'
