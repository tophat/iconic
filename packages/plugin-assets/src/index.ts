import * as path from 'path'

import * as fs from 'fs-extra'
import * as glob from 'glob'

import {
    NamingScheme,
    Plugin,
    PluginParams,
} from '@react-vector-graphics/types'

import { pathToName } from './utils'

const ns = (s: TemplateStringsArray): string => `assets/${s}`

export const OPTIONS = {
    FILE_EXT: ns`fileExt`,
    GLOB_PATTERN: ns`globPattern`,
    NAME_SCHEME: ns`nameScheme`,
    OUTPUT_PATH: ns`outputPath`,
}

const run: Plugin = async (code, config, state) => {
    if (code) {
        const outputPath = config.options[OPTIONS.OUTPUT_PATH]
        if (outputPath && state.componentName) {
            const pathToFile = path.join(outputPath, state.componentName)
            const fileExt = config.options[OPTIONS.FILE_EXT]
            const filePath = fileExt ? `${pathToFile}.${fileExt}` : pathToFile
            fs.outputFileSync(filePath, code)
        }
        return { code, state }
    } else {
        const globPattern: string = config.options[OPTIONS.GLOB_PATTERN]
        const nameScheme: NamingScheme = config.options[OPTIONS.NAME_SCHEME]
        return glob.sync(globPattern).map(
            (filePath: string): PluginParams => ({
                code: fs.readFileSync(filePath, { encoding: 'utf-8' }),
                state: Object.assign(
                    {
                        componentName: pathToName(filePath, nameScheme),
                        filePath,
                    },
                    state,
                ),
            }),
        )
    }
}

export default run
