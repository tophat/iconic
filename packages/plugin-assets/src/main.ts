import * as path from 'path'

import * as fs from 'fs-extra'
import * as glob from 'glob'

import {
    Logger,
    NamingScheme,
    Plugin,
    PluginParams,
    State,
} from '@react-vector-graphics/types'

import { pathToName } from './utils'
import { STATE } from './constants'

import { OPTIONS } from '.'

const findAssets = (args: {
    globPattern: string
    nameScheme: NamingScheme
    state: State
}): PluginParams[] => {
    return glob.sync(args.globPattern).map(
        (filePath: string): PluginParams => ({
            code: fs.readFileSync(filePath, { encoding: 'utf-8' }),
            state: Object.assign(
                {
                    [STATE.COMPONENT_NAME]: pathToName(
                        filePath,
                        args.nameScheme,
                    ),
                    [STATE.FILE_PATH]: filePath,
                },
                args.state,
            ),
        }),
    )
}

const writeComponent = (args: {
    assetFile?: string
    code: string
    componentName?: string
    componentFiles: { [fileName: string]: string }
    outputPath?: string
    fileExt?: string
    logger: Logger
}): void => {
    if (!args.componentName) {
        return args.logger.warn(
            `No '${STATE.COMPONENT_NAME}' provided for '${args.assetFile}'.`,
        )
    }
    if (!args.outputPath) {
        return args.logger.warn(`No '${OPTIONS.OUTPUT_PATH}' provided.`)
    }
    if (!args.fileExt) {
        args.logger.warn(`No '${OPTIONS.FILE_EXT}' provided.`)
    }

    const componentFiles = Object.entries(args.componentFiles)
    const pathToFolder = path.join(
        args.outputPath,
        componentFiles.length ? args.componentName : '',
    )
    const pathToFile = path.join(
        pathToFolder,
        componentFiles.length ? 'index' : args.componentName,
    )
    const componentFilePath = args.fileExt
        ? `${pathToFile}.${args.fileExt}`
        : pathToFile
    fs.outputFileSync(componentFilePath, args.code)
    for (const [fileName, fileContents] of componentFiles) {
        const filePath = path.join(pathToFolder, fileName)
        fs.outputFileSync(filePath, fileContents)
    }
}

export const run: Plugin = async (code, config, state, logger) => {
    if (code) {
        writeComponent({
            assetFile: state[STATE.FILE_PATH],
            code,
            componentFiles: state[STATE.COMPONENT_FILES] ?? {},
            componentName: state[STATE.COMPONENT_NAME],
            fileExt: config.options[OPTIONS.FILE_EXT],
            logger,
            outputPath: config.options[OPTIONS.OUTPUT_PATH],
        })
        return { code, state }
    } else {
        return findAssets({
            globPattern: config.options[OPTIONS.GLOB_PATTERN],
            nameScheme: config.options[OPTIONS.NAME_SCHEME],
            state,
        })
    }
}
