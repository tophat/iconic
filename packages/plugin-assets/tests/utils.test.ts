import { pathToName } from '../src/utils'

describe('plugin-assets/utils', () => {
    it.each`
        nameScheme
        ${`CONSTANT_CASE`}
        ${`camelCase`}
        ${`PascalCase`}
        ${`snake_case`}
        ${`spinal-case`}
    `(
        'converts file path to correct naming scheme $nameScheme',
        ({ nameScheme }) => {
            expect(
                pathToName(
                    'example/assets/file3_white.24x24.icon.svg',
                    nameScheme,
                ),
            ).toMatchSnapshot()
        },
    )
})
