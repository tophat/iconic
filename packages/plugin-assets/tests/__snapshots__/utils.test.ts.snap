// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-assets/utils converts file path to correct naming scheme CONSTANT_CASE 1`] = `"FILE3_WHITE_24X24_ICON"`;

exports[`plugin-assets/utils converts file path to correct naming scheme PascalCase 1`] = `"File3White24x24Icon"`;

exports[`plugin-assets/utils converts file path to correct naming scheme camelCase 1`] = `"file3White24x24Icon"`;

exports[`plugin-assets/utils converts file path to correct naming scheme snake_case 1`] = `"file3_white_24x24_icon"`;

exports[`plugin-assets/utils converts file path to correct naming scheme spinal-case 1`] = `"file3-white-24x24-icon"`;
